python/cudf/dataframe/column.py

```diff
@@ -446,29 +456,63 @@ def __getitem__(self, arg):
 
-    def masked_assign(self, value, mask):
-        """Assign a scalar value to a series using a boolean mask
-        df[df < 0] = 0
-
-        Parameters
-        ----------
-        value : scalar
-            scalar value for assignment
-        mask : cudf Series
-            Boolean Series
-
-        Returns
-        -------
-        cudf Series
-            cudf series with new value set to where mask is True
+    # def masked_assign(self, value, mask):
+    #     """Assign a scalar value to a series using a boolean mask
+    #     df[df < 0] = 0
+
+    #     Parameters
+    #     ----------
+    #     value : scalar
+    #         scalar value for assignment
+    #     mask : cudf Series
+    #         Boolean Series
+
+    #     Returns
+    #     -------
+    #     cudf Series
+    #         cudf series with new value set to where mask is True
+    #     """
+
+    #     # need to invert to properly use gpu_fill_mask
+    #     mask_invert = mask._column._invert()
+    #     out = cudautils.fill_mask(data=self.data.to_gpu_array(),
+    #                               mask=mask_invert.as_mask(),
+    #                               value=value)
+    #     return self.replace(data=Buffer(out), mask=None, null_count=0)
+
+    def masked_assign(self, data, mask, isnull):
+        """Assign values in `data` to this column, subject to `mask`
         """
+        def _to_devarray(arg, name):
+            try:
+                return rmm.to_device(arg)
+            except Exception:
+                raise TypeError(
+                    f"Unsupported type {type(arg)} for argument {name}"
+                )
 
-        # need to invert to properly use gpu_fill_mask
-        mask_invert = mask._column._invert()
-        out = cudautils.fill_mask(data=self.data.to_gpu_array(),
-                                  mask=mask_invert.as_mask(),
-                                  value=value)
-        return self.replace(data=Buffer(out), mask=None, null_count=0)
+        if isinstance(data, Number):
+            data = data * np.ones(len(mask))
+        data = _to_devarray(data, "data")
+        mask = _to_devarray(mask, "mask")
+
+        if data.size != mask.size:
+            raise ValueError(
+                "Arguments should have the same size "
+                f"(found {data.size} and {mask.size})"
+            )
+        for i, val in enumerate(isnull):
+            if mask[i]:
+                nullmap_idx, pos = divmod(i, 8)
+                bitmask = 1 << pos
+                if not val:
+                    self.nullmask.mem[nullmap_idx] = self.nullmask.mem[nullmap_idx] | bitmask
+                else:
+                    self.nullmask.mem[nullmap_idx] = self.nullmask.mem[nullmap_idx] & ~bitmask
+
+        threads_per_block = 32
+        blocks_per_grid = (data.size + (threads_per_block - 1)) // threads_per_block
+        boolean_assign_kernel[blocks_per_grid, threads_per_block](data, mask, self.data.mem)
```

Comment:
Instead of calculating these all (unless there's a strong reason), you can do `boolean_assign_kernel.forall(data.size)(data, mask, self.data.mem)` which automatically calculates these nicely for you.